xds/src/main/java/io/grpc/xds/AdsStream.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.Status;
+import io.grpc.stub.StreamObserver;
+import io.grpc.xds.shaded.envoy.api.v2.DiscoveryRequest;
+import io.grpc.xds.shaded.envoy.api.v2.DiscoveryResponse;
+import io.grpc.xds.shaded.envoy.service.discovery.v2.AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub;
+
+/**
+ * Ads client implementation.
+ */
+final class AdsStream implements StreamObserver<DiscoveryResponse> {
+  private final AggregatedDiscoveryServiceStub stub;
+
+  private StreamObserver<DiscoveryRequest> xdsRequestWriter;
+  private boolean closed;
+
+  AdsStream(AggregatedDiscoveryServiceStub stub) {
+    this.stub = checkNotNull(stub, "stub");
+  }
+
+  void start() {
+    xdsRequestWriter = stub.withWaitForReady().streamAggregatedResources(this);
+  }
+
+  @Override
+  public void onNext(DiscoveryResponse value) {
+    // TODO: impl
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    // TODO: impl
+  }
+
+  @Override
+  public void onCompleted() {
+    // TODO: impl
+  }
+
+  void close(Status status) {
+    if (closed) {
+      return;
+    }
+    closed = true;
```

Comment:
Should this be synchronized? The `onError` doc says it may be called only once.

Reply:
This will only be called in a `SynchronizationContext` and no extra synchronization is needed, in the same pattern as the other lb implementations. I will add comments on the thread-safety guidance on all classes once they are close to fully baked.